### PR TITLE
feat(webdav): add bearer_token_command for dynamic token acquisition

### PR DESCRIPTION
**Pull request by GreenHatHG**
see: https://github.com/iterative/dvc/pull/10917

### CODE REVIEW

File: dvc_webdav/bearer_auth_client.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        # DVC might catch this and swallow it, but we've done our duty to notify.
+        raise
+
+    token = result.stdout.strip()
+    if not token:
+        raise ValueError("Command executed successfully but returned an empty token.")
+    return token
+
+
+class BearerAuthClient(httpx.Client):
```
Comment (skshetry) [inline dvc_webdav/bearer_auth_client.py]: Do we need to extend `httpx.Client`? Could we make this work with custom [`httpx.Auth`](https://www.python-httpx.org/advanced/authentication/)?

**Review by skshetry — APPROVED** [review]
Thank you for contributing. This looks good overall. One question. Could you also add a few tests to exercise the `BearerAuth` code path? That would help ensure it works correctly and does not regress.

Also, contribution to dvc repositories now require accepting the CLA: https://cla-assistant.io/treeverse/dvc-webdav. It'd be great if you could do that. Thanks.

File: dvc_webdav/bearer_auth.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        if response.status_code == 401:
+            _log_with_thread(
+                logging.DEBUG, "[BearerAuth] Received 401. Attempting recovery."
+            )
+
+            # Read response body to release connection
+            try:
+                response.read()
+            except Exception as e:  # noqa: BLE001
+                logger.debug("Error reading response body: %s", e)
```
[the remark this answers is not in the document]
Comment (GreenHatHG) [reply dvc_webdav/bearer_auth.py]: Fixed in 773944a.
I verified that removing `response.read()` works correctly.

**Comment by GreenHatHG** [comment]
@skshetry 
I've added the test cases covering `BearerAuth` as requested.
Also, to ensure end-to-end stability, I've added a corresponding integration test in the dvc repo (treeverse/dvc#10917).

File: dvc_webdav/tests/test_bearer_auth.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+    req2 = flow.send(response_401)
+
+    # Server returns 401 AGAIN (for the new token-v2)
+    resp_401_again = httpx.Response(401, request=req2)
+
+    # Generator must terminate (raising StopIteration),
+    # indicating no further retries will be attempted.
+    with pytest.raises(StopIteration):
+        flow.send(resp_401_again)
+
```
Comment (skshetry) [inline dvc_webdav/tests/test_bearer_auth.py]: I don’t see the need for tests beyond this line, especially the concurrent one. I doubt they would help catch the issue if it occurs in the future.

When I requested tests, my only concern was verifying that the command is called and behaves as expected: setting the token in the header on success and raising an error when it exits with a failure. Another goal was to ensure it works on Windows, though the current test doesn’t cover that since all subprocess calls are mocked and I’m okay with that.

I appreciate that you also added tests for token refresh, but the rest feels excessive.

Comment (GreenHatHG) [reply dvc_webdav/tests/test_bearer_auth.py]: Thanks for the feedback!
I agree that the test suite was getting a bit heavy. I've removed `test_auth_flow_failure_after_refresh` and the other edge-case tests as suggested.
It looks much cleaner now.  Please let me know if there's anything else needed.

**Comment by GreenHatHG** [comment]
@skshetry 
It seems the GitHub Actions workflow is waiting for approval to run. Could you please hit the "Approve and run" button when you have a moment? Thanks!
图片

**Comment by skshetry** [comment]
@GreenHatHG, looks like we don't have `pytest-mock` installed here. Please feel add them to `tests` dependencies in pyproject.toml.

**Comment by GreenHatHG** [comment]
> @GreenHatHG, looks like we don't have `pytest-mock` installed here. Please feel add them to `tests` dependencies in pyproject.toml.

Fixed in the latest commit. 
All CI checks have passed. The remaining conversation threads are addressed in the code.
I think this is ready to be merged whenever you have a moment.

**Comment by skshetry** [comment]
Thank you for contributing!
